File: backend/src/main/java/org/ucjgsalm/locagest/repository/LigneSejourRepository.java
package org.ucjgsalm.locagest.repository;

import jakarta.inject.Singleton;
import javax.sql.DataSource;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;
import org.ucjgsalm.locagest.domain.LigneSejour;
import org.ucjgsalm.locagest.domain.enums.*;

@Singleton
public class LigneSejourRepository {

    private final DataSource ds;

    public LigneSejourRepository(DataSource ds) { this.ds = ds; }

    public List<LigneSejour> findBySejourId(UUID sejourId) throws SQLException {
        String sql = """
            SELECT * FROM ligne_sejour
            WHERE sejour_id = ?
            ORDER BY type_ligne, created_at
            """;
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement(sql)) {
            ps.setObject(1, sejourId);
            return mapAll(ps.executeQuery());
        }
    }

    /** Toutes les lignes LIBRE de tous les séjours — onglet "À promouvoir" du trésorier/resp. */
    public List<LigneSejour> findAllLibre() throws SQLException {
        String sql = """
            SELECT ls.*, s.date_arrivee, l.nom AS loc_nom
            FROM ligne_sejour ls
            JOIN sejour s ON s.id = ls.sejour_id
            JOIN locataire l ON l.id = s.locataire_id
            WHERE ls.statut = 'LIBRE'
            ORDER BY ls.created_at DESC
            """;
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement(sql)) {
            return mapAll(ps.executeQuery());
        }
    }

    public Optional<LigneSejour> findById(UUID id) throws SQLException {
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement("SELECT * FROM ligne_sejour WHERE id = ?")) {
            ps.setObject(1, id);
            var rs = ps.executeQuery();
            if (rs.next()) return Optional.of(map(rs));
            return Optional.empty();
        }
    }

    /** Insère une ligne. Ne pas passer montant (GENERATED). */
    public LigneSejour insert(LigneSejour l) throws SQLException {
        String sql = """
            INSERT INTO ligne_sejour
                (id, sejour_id, config_item_id, type_ligne, designation,
                 quantite, prix_unitaire, statut, saisi_par)
            VALUES (?, ?, ?, ?::type_ligne, ?, ?, ?, ?::statut_ligne, ?)
            RETURNING *
            """;
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement(sql)) {
            ps.setObject(1, l.id());
            ps.setObject(2, l.sejourId());
            ps.setObject(3, l.configItemId());
            ps.setString(4, l.typeLigne().name());
            ps.setString(5, l.designation());
            ps.setBigDecimal(6, l.quantite());
            ps.setBigDecimal(7, l.prixUnitaire());
            ps.setString(8, l.statut().name());
            ps.setString(9, l.saisiPar());
            var rs = ps.executeQuery();
            rs.next();
            return map(rs);
        }
    }

    /**
     * Supprime toutes les lignes d'un type donné pour un séjour.
     * Doit être appelée dans un contexte {@code @Transactional}.
     */
    public void deleteBySejourIdAndType(UUID sejourId, TypeLigne type) throws SQLException {
        String sql = "DELETE FROM ligne_sejour WHERE sejour_id = ? AND type_ligne = ?::type_ligne";
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement(sql)) {
            ps.setObject(1, sejourId);
            ps.setString(2, type.name());
            ps.executeUpdate();
        }
    }

    /**
     * Promotion LIBRE → SUPPLEMENT :
     * met à jour type_ligne, statut et config_item_id de la ligne existante.
     */
    public void promouvoir(UUID ligneId, UUID configItemId) throws SQLException {
        String sql = """
            UPDATE ligne_sejour
            SET type_ligne     = 'SUPPLEMENT'::type_ligne,
                statut         = 'CONFIRME'::statut_ligne,
                config_item_id = ?
            WHERE id = ? AND statut = 'LIBRE'
            """;
        try (var conn = ds.getConnection();
             var ps = conn.prepareStatement(sql)) {
            ps.setObject(1, configItemId);
            ps.setObject(2, ligneId);
            int updated = ps.executeUpdate();
            if (updated == 0) throw new IllegalStateException(
                "Ligne " + ligneId + " introuvable ou déjà promue");
        }
    }

    // ── Mapping ────────────────────────────────────────────────────────────

    private LigneSejour map(ResultSet rs) throws SQLException {
        return new LigneSejour(
            rs.getObject("id",             UUID.class),
            rs.getObject("sejour_id",      UUID.class),
            rs.getObject("config_item_id", UUID.class),
            TypeLigne.valueOf(rs.getString("type_ligne")),
            rs.getString("designation"),
            rs.getBigDecimal("quantite"),
            rs.getBigDecimal("prix_unitaire"),
            rs.getBigDecimal("montant"),
            StatutLigne.valueOf(rs.getString("statut")),
            rs.getString("saisi_par"),
            rs.getObject("created_at", java.time.OffsetDateTime.class)
        );
    }

    private List<LigneSejour> mapAll(ResultSet rs) throws SQLException {
        var list = new ArrayList<LigneSejour>();
        while (rs.next()) list.add(map(rs));
        return list;
    }
}

File: plugin/includes/Repository/LigneSejourRepository.php
<?php

declare(strict_types=1);

namespace Locagest\Repository;

class LigneSejourRepository {

    private string $table;

    public function __construct() {
        global $wpdb;
        $this->table = $wpdb->prefix . 'locagest_ligne_sejour';
    }

    public function find_by_sejour( int $sejour_id ): array {
        global $wpdb;
        return $wpdb->get_results(
            $wpdb->prepare( "SELECT * FROM {$this->table} WHERE sejour_id = %d ORDER BY type_ligne ASC, ordre ASC", $sejour_id ),
            ARRAY_A
        ) ?: [];
    }

    public function find_lignes_libres_en_attente(): array {
        global $wpdb;
        return $wpdb->get_results(
            "SELECT l.*, s.date_debut, s.date_fin FROM {$this->table} l
             JOIN {$wpdb->prefix}locagest_sejour s ON s.id = l.sejour_id
             WHERE l.type_ligne = 'LIBRE' AND l.statut = 'BROUILLON'
             ORDER BY l.created_at DESC",
            ARRAY_A
        ) ?: [];
    }

    public function find_by_id( int $id ): ?array {
        global $wpdb;
        return $wpdb->get_row(
            $wpdb->prepare( "SELECT * FROM {$this->table} WHERE id = %d", $id ),
            ARRAY_A
        ) ?: null;
    }

    public function create( array $data ): int {
        global $wpdb;
        $wpdb->insert( $this->table, $data );
        return (int) $wpdb->insert_id;
    }

    public function update( int $id, array $data ): void {
        global $wpdb;
        $wpdb->update( $this->table, $data, [ 'id' => $id ] );
    }

    /** Supprime toutes les lignes d'un type donné pour un séjour (pour recalcul). */
    public function delete_by_sejour_and_type( int $sejour_id, string $type_ligne ): void {
        global $wpdb;
        $wpdb->query( $wpdb->prepare(
            "DELETE FROM {$this->table} WHERE sejour_id = %d AND type_ligne = %s",
            $sejour_id,
            $type_ligne
        ) );
    }

    /** Supprime toutes les lignes calculées d'un séjour (HEBERGEMENT, ENERGIE, TAXE). */
    public function delete_calculated_lines( int $sejour_id ): void {
        global $wpdb;
        $wpdb->query( $wpdb->prepare(
            "DELETE FROM {$this->table} WHERE sejour_id = %d AND type_ligne IN ('HEBERGEMENT','ENERGIE','TAXE')",
            $sejour_id
        ) );
    }

    /** Retourne la première ligne d'un séjour correspondant à un config_item donné. */
    public function find_by_sejour_and_config_item( int $sejour_id, int $config_item_id ): ?array {
        global $wpdb;
        return $wpdb->get_row(
            $wpdb->prepare(
                "SELECT * FROM {$this->table} WHERE sejour_id = %d AND config_item_id = %d LIMIT 1",
                $sejour_id,
                $config_item_id
            ),
            ARRAY_A
        ) ?: null;
    }

    /** Promeut une ligne LIBRE en SUPPLEMENT avec un config_item_id. */
    public function promouvoir( int $ligne_id, int $config_item_id ): void {
        global $wpdb;
        $wpdb->update( $this->table, [
            'type_ligne'     => 'SUPPLEMENT',
            'config_item_id' => $config_item_id,
            'statut'         => 'CONFIRME',
        ], [ 'id' => $ligne_id ] );
    }
}
